File: 0073-Set-Matrix-Zeroes/0073-Set-Matrix-Zeroes.py
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None. Do not return anything, modify matrix in-place.
        """
        rows = len(matrix)
        cols = len(matrix[0])

        zero_rows = set()  # store row indices that need to be zero
        zero_cols = set()  # store col indices that need to be zero

        # Step 1: Identify rows and cols to be zeroed
        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == 0:
                    zero_rows.add(i)
                    zero_cols.add(j)

        # Step 2: Set the cells to zero
        for i in range(rows):
            for j in range(cols):
                if i in zero_rows or j in zero_cols:
                    matrix[i][j] = 0

File: 0073-Set-Matrix-Zeroes/0073-Set-Matrix-Zeroes.c
#include <stdbool.h>  // For using bool type
#include <stdlib.h>   // For malloc and free

void setZeroes(int** matrix, int matrixSize, int* matrixColSize) {
    int rows = matrixSize;
    int cols = matrixColSize[0];

    // Step 1: Create markers for which rows/cols to zero
    bool* zero_rows = (bool*)calloc(rows, sizeof(bool));  // all false (0)
    bool* zero_cols = (bool*)calloc(cols, sizeof(bool));  // all false (0)

    // Step 2: Scan matrix and mark rows and cols
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (matrix[i][j] == 0) {
                zero_rows[i] = true;  // mark this row
                zero_cols[j] = true;  // mark this col
            }
        }
    }

    // Step 3: Second pass — update the matrix in-place
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (zero_rows[i] || zero_cols[j]) {
                matrix[i][j] = 0;
            }
        }
    }

    // Step 4: Free memory
    free(zero_rows);
    free(zero_cols);
}

